File: LeetCode/200.number-of-islands.java
/*
 * @lc app=leetcode id=200 lang=java
 *
 * [200] Number of Islands
 */

import java.util.*;

// @lc code=start

class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0; // handle null or empty grid
        }

        boolean[][] visited = new boolean[grid.length][grid[0].length];

        int[] islandCount = new int[1];

        for (int row = 0; row < grid.length; row++) {
            for (int col = 0; col < grid[0].length; col++) {

                if (grid[row][col] == '1' && visited[row][col] != true) {
                    this.checkAllConnectedLand(grid, visited, row, col);
                    islandCount[0]++;
                }

            }
        }
        return islandCount[0];
    }

    private void checkAllConnectedLand(char[][] grid, boolean[][] visited, int row, int col) {
        if (isOutOfBound(grid, row, col) || visited[row][col] || grid[row][col] == '0')
            return;

        visited[row][col] = true;

        checkAllConnectedLand(grid, visited, row - 1, col);
        checkAllConnectedLand(grid, visited, row + 1, col);
        checkAllConnectedLand(grid, visited, row, col - 1);
        checkAllConnectedLand(grid, visited, row, col + 1);
    }

    private boolean isOutOfBound(char[][] grid, int row, int col) {
        return row < 0 || row >= grid.length || col < 0 || col >= grid[0].length;
    }
}

// @lc code=end

File: LeetCode/200.number-of-islands.py
#
# @lc app=leetcode id=200 lang=python3
#
# [200] Number of Islands
#

from typing import List

# @lc code=start
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        memo: List[List[bool]] = [[False for x in range(len(grid[0]))] for _ in range(len(grid))]
        count = 0
        for rowIdx, row in enumerate(grid):
            for colIdx, col in enumerate(row):
                if col == "1" and  notmemo[rowIdx][colIdx] :
                    self.dfs(grid, rowIdx, colIdx, memo)
                    count += 1
        return count
        
    def dfs(self, grid: List[List[str]], row: int, col: int, memo: List[List[bool]]):
        maxRow = len(grid) - 1
        maxCol = len(grid[0]) - 1
        
        # if out of bound
        if row < 0 or col < 0 or row > maxRow or col > maxCol:
            return False
        
        if grid[row][col] == "0":
            return False
        
        # if visited
        if memo[row][col]:
            return False
        
        memo[row][col] = True

        self.dfs(grid, row - 1, col, memo)
        self.dfs(grid, row + 1, col, memo)
        self.dfs(grid, row, col - 1, memo)
        self.dfs(grid, row, col + 1, memo)


# @lc code=end

